### ISSUE TITLE
One Key binding that auto detects formatter to use

### ISSUE THREAD

**Issue by bhuynhdev** [bitst0rm-pub/Formatter#40]
Is there a way to add one key binding that will run a command that does auto detection of what formatters to use.

The currently provided example key bindings json use the "run_format" command name with different args for different formatter. Is there a way or a command so that with only one key binding but the formatting still works across different languages?

**Comment by bitst0rm** [comment]
Hi,

The challenge is that it is not possible to predict which languages and formatters a user prefers to use. Out of over 20 plugins, 15 are capable of formatting CSS, for example. In such a scenario, Formatter cannot determine which specific one out of the 15 it should use for the CSS job. It requires explicit an preferred plugin ID for each key binding.
Therefore, a single key binding for all across different languages would not work well. If, then it would be somewhat heuristic in nature. A similar situation and its solution has been mentioned in the "format_on_save" settings.

An alternative approach could be to introduce a new command allowing users to define plugins to bind to a single key:
```
"format_on_what": ["plugin_a", "plugin_m"] // a single key binding for all plugins in the array
```

The problem: if both "plugin_a" and "plugin_m" also supports CSS. In this case, we will find ourselves in the circular situation mentioned above.

**Comment by bhuynhdev** [comment]
Hi. Thanks for the lightning fast reply!

Would it be possible to rely on the `syntaxes` field in the Formatter plugin settings?

Then it becomes user's responsibility to specify which formatter to use for which syntax. If multiple formatter conflicts for a syntax, then either errors out, or ask users to choose a default formatter.

**Comment by midrare** [comment]
Letting the user specify a preferred formatter per language in the user config file is what people have been doing in Neovim land for years. But a better idea might be to read a user-specified formatter from the sublime-project file, or travel the dir tree upwards from the current file looking for a `.formatter` file. This way you can auto-conform to project guidelines when you switch between repos.

**Comment by bitst0rm** [comment]
Hi @midrare,

There is already a function in place for traversing the tree upwards. However, attempting to use a per-project file `.formatter `with just one command (with or without a single key binding) doesn't yield the expected result due to the following reasons:

- Formatter requires an assigned key type for a specific formatter to function correctly (ok, can be hacked).
- When iterating through multiple formatters defined in `.formatter`, it results in looping through all of them:
  - looping through multiple selections again (selections format).
  - it recurses through the folder format once more (recursive folders format).
  - to determine when the initial loop finished to initiate the subsequent loop etc.
 
In other words, it would work for a simple uni concept, but not for multi formatter as i see.

**Comment by midrare** [comment]
@bitst0rm I'm not sure what you mean by "looping through all of them".

My idea is that every `.formatter` file is a file type-to-formatter map. For example:

```json
{
  "*.cpp": "clang-format",
  "*.py": "yapf",
}
```

When the user requests a file to be formatted, you'd (starting with the file's containing dir) travel up the file tree looking for `.formatter` files and merge them into a single map, with files in parent folders overridden by those in child folders, like how `.gitignore` works. Then you use the merged map to determine what formatter to use to format the file. (If a map is defined in the `sublime-project` file it's equivalent to a `.formatter` file in the project's root dir.) If you've used *vim-projectionist* this is basically how it works.

Maybe you had something different in mind, or maybe I'm misunderstanding something?

**Comment by bitst0rm** [comment]
@midrare 
I mean if `.formatter` content is something like that one in `Formatter.sublime-settings` with its huge keys and values to control Formatter like `"recursive_folder_format"`, then we would end in an infinite loop.

Your idea for simplifying the structure of `.formatter` sounds promising. Based on it, i'm currently testing to implement this feature. Patched 4 files, and initial tests look promising, but I need more time to ensure a proper implementation.
At the moment I cannot say or promise, if it will end with happy-end or just a time-wasting.

This is the structure of the test file `.formatter`:
```json
{
  "autopep8": {
     "syntaxes": ["jkedkkjsjhj", "python", "gfhfkdfkkdf"],
     "exclude_syntaxes": {
         "html": ["markdown"],
         "all": ["markdown"]
     }
  },
  "jsbeautifier": {
     "syntaxes": ["awesrdfvtzug", "js", "css", "html", "dfgvhbioj"],
     "exclude_syntaxes": {
         "html": ["markdown"],
         "all": ["markdown"]
     }
  }
}
```

**Comment by bitst0rm** [comment]
@midrare
ok, it is ready for test in the branch [formatter_project_config](https://github.com/bitst0rm-pub/Formatter/tree/formatter_project_config)
- one-command, one-keybinding
- Main menu and context menu include both a new item: `"Project Format File"`
  - This new item only appears if a project file found. (as convenient indicator)
  - I need a visible file, so for test only `formatter.json` or  `.formatter` as project file is accepted. (later it will be dot files `.formatter.json`, `.formatter`)
 
Please test it to give feedback or maybe I'm misunderstanding.
And `"Project Format File"` as name, ok? or `"Format File Project"` or your suggestions?

_formatter.json_ structure
```json
{
  "config": {
     "format_on_save": false,
     "format_on_paste": false,
  },
  "jsbeautifier": {
     "syntaxes": ["awesrdfvtzug", "js", "css", "html", "dfgvhbioj"],
     "exclude_syntaxes": {
         "html": ["markdown"]
     }
  },
  "autopep8": {
     "syntaxes": ["jkedkkjsjhj", "python", "gfhfkdfkkdf"],
  }
}
```

**Comment by midrare** [comment]
@bitst0rm Functionality-wise everything is looking good. No problems so far on my end.

---

I have a few thoughts about the structure of `formatter.json`.

 - **Syntax-to-formatter instead of formatter-to-syntax**
   So for example `{ "js": "jsbeautifier" }` instead of `{ "jsbeautifier": { "syntaxes": ["js"] }`. There are a couple advantages to this.
   - You can see just by looking at the file that all the syntaxes you need are supported. IMO seeing at-a-glance which syntaxes are supported is more useful than seeing at-a-glance which formatters are active.
   - You can guarantee that only one correct formatter assigned for each syntax.
   
     In a formatter-to-syntax map, if two formatters support the same syntax, which one takes precedence?

     ```json
     {
       "autopep8": { "syntaxes": ["python"] },
       "yapf": { "syntaxes": ["python"] }
     }
	 ```
	
     In a syntax-to-formatter map, this problem becomes impossible.

     ```json
     {
       "python": "autopep8"
     }
	 ```

 - **No config options in `formatter.json`**
   If user-specific config options (i.e. `format_on_save`, `format_on_paste` are included in `formatter.json`, then you can't upload it to a public repo since every user's prefs are different. It could be split it into `formatter.user.json`. Then `formatter.json` can be uploaded publicly and `formatter.user.json` added to `.gitignore`.
 - **New name for `formatter.json`? `sublimeformatter.json`?**
   `formatter.json` sounds a little generic. Someone looking at this wouldn't recognize it as a file specific to the Sublime Text Formatter plugin. They might think it has effects on the build process or a git hook, make inappropriate edits, etc. Also it `formatter.json` would be impossible to google so they'd never find out what it's actually for.

---

<sub>I'm getting a weird error when cloning the repo. I had to use "Download as ZIP" to test. Any tips?</sub>

<details>
<summary><sub>Error output</sub></summary>

```bash
C:\Users\midrare\AppData\Roaming\Sublime Text\Packages>git clone git@github.com:bitst0rm-pub/Formatter.git
Cloning into 'Formatter'...
remote: Enumerating objects: 4056, done.
remote: Counting objects: 100% (965/965), done.
remote: Compressing objects: 100% (268/268), done.
remote: Total 4056 (delta 876), reused 747 (delta 696), pack-reused 3091
Receiving objects: 100% (4056/4056), 1.13 MiB | 2.44 MiB/s, done.
Resolving deltas: 100% (3188/3188), done.

C:\Users\midrare\AppData\Roaming\Sublime Text\Packages>cd Formatter

C:\Users\Monk\AppData\Roaming\Sublime Text\Packages\Formatter>git status
fatal: not a git repository (or any of the parent directories): .git
```
</details>

**Comment by bitst0rm** [comment]
@midrare 
Thanks for your suggestions and tests. I have pushed the final implementation to the `master` branch with an updated [Readme](https://github.com/bitst0rm-pub/Formatter?tab=readme-ov-file#per-project-configuration). It is based on your latest suggestion with Syntax-to-formatter structure.

If everything ok, then i will release it as 1.4.0

If you clone Formatter and run it inside Sublime, then the `.git` of Formatter does not exist anymore as it has been auto removed. This folder is the main trouble causing a lot of issue and conflicts when i updated using sublime 4 with its build-in git merge functionality.

**Comment by midrare** [comment]
@bitst0rm  `master` branch is working for me 👍

A bit late, but maybe rename the command to something like `Formatter: Beautify (auto-detect)`? I'm purposely excluding any mention of "project" because of an idea for extending auto-detect. `Formatter.sublime-settings` could include an embedded `.sublimeformatter` structure which would serve as the default `.sublimeformatter` (effectively equivalent to a `.sublimeformatter` in the root drive). This way auto-format can be both in and outside of projects, and it will automatically do the right thing.

**Comment by bitst0rm** [comment]
@midrare 
All your ideas and suggestions were gold value, it is not late. Thanks! 👍
You are right, using the term "project" is a bit ambitious. I have changed from `Project Format File` to `Auto Format File` as `Auto-detect` is a bit long for the context menu. I will see to integrate this feature inside `Formatter.sublime-settings`, so it would work in and outside.

**Comment by bitst0rm** [comment]
@midrare 
It is done 2/3 on my end as draft, but i still need your suggestions as 2. opinion.
1. Naming:
  - `Auto Format File` (we already have this for the outside)
  - `Auto Format File` (I need a name for the inside. Currently, i'm considering `Auto Format File Internal`, but it feels lengthy and imprecise. Any suggestions?)
  - Or only one `Auto Format File` (with one-command/keybinding) for both cases with priority:
    - outside first, inside last
    - merge 2 in one, outside overwrites inside
 2. Issue:
   - In cases where both the inside and outside versions coexist, which should take priority? The design principle of Formatter is always from the user perspective: `Formatter.sublime-settings` first (=reflecting the user's wishes) and local plugins last (=reflecting external wishes). But in this case, it could be the opposite if a user put (=wish) the `sublimeformatter.json` to use. I think `sublimeformatter.json` should be priorized. This issue is the case where `format_on_save/paste` comes into play, which I would like to add it into structure as something like below because `format_on_save/paste` at `auto_format` and `"project context"` level does not have to do anymore with that one from 3rd-party level (no effect for `format_on_save/paste` at 3rd-party level if `auto_format` or `"project context"` is set)
  3. Structure:
```js
{
    "auto_format": { // Should it be placed here? i think better here.
        "config": {
            "format_on_save": false,
            "format_on_paste": false,
        },
        "json": {
            "uid": "jsbeautifier",
            "exclude_syntaxes": {
                "html": ["markdown"]
            }
        },
        "python": {
            "uid": "autopep8"
        }
    },
    // THIRD-PARTY PLUGINS LEVEL
    "formatters": {
         "auto_format": {} // Or should it be placed here?
         "plugin_a": {},
         "plugin_b": {},
     }
}
```

**Comment by midrare** [comment]
@bitst0rm

 1. Naming: just one `Auto Format File` (explanation below)
 2. Priority: `.sublimeformatter.json` takes priority (explanation below)
 3. `Formatter.sublime-settings` structure: I agree, `auto_format` should be separate.

---

Ideally there would just be one option, `Auto Format File`, with bottom-up override.

So it would be like this:

  1. User calls `Auto Format File` on a file.
  2. First, read syntax-to-formatter map from `Formatter.sublime-settings`
  3. Merge `helloworld/.sublimeformatter`, letting the new map overwrite the old map.
  4. Merge `helloworld/util/.sublimeformatter`, letting the new map overwrite the old map.
  5. Keep recursing down the file tree, overwriting the parent maps with child maps
  6. Use the final merged map to determine the right formatter to use, and finally run the formatter on the file.

---

I don't think top-down override (as opposed to bottom-up override) is worth it. There's no way to use top-down override without creating flexibility problems.

Lets say you have this project:

```
    helloworld/
    ├── util/
    |   ├── .sublimeformatter.json
    |   └── sub.py
    ├── .sublimeformatter.json
    └── main.py
```

Now you have two situations:

 1. If the user wants both `main.py` and `util/sub.py` with `autopep8`
    - Bottom-up override: the user can do it by deleting `util/.sublimeformatter.json`. It will inherit `autopep8` from the parent dir's `.sublimeformatter.json`
    - Top-down override: the user can do it by setting `.sublimeformatter.json` to `autopep8` in the parent dir.
 2. If the user wants `main.py` with `autopep8` but `util/sub.py` with `yapf`
    - Bottom-up override: the user can do it by setting `autopep8` in `.sublimeformatter.json` and setting `yapf` in `util/.sublimeformatter.json`.
    - Top-down override: the user has no choice but to format one of the files incorrectly. If `.sublimeformatter.json` is set to `autopep8` it will override `util/.sublimeformatter.json` and `util/sub.py` will be wrongly formatted with `autopep8`. The only way to fix this is to delete `.sublimeformatter.json`, but then `main.py` will have no formatter. So this problem is impossible to fix.

So, bottom-up can handle both situation 1 and 2, but top-down can only handle situation 1.

I don't think there needs to be a top-down override at all, even as an option. The one advantage to top-down override is if a `.sublimeformatter.json` file is counter to what the user wants, but in practice (from experience with Neovim plugins) I've never seen this happen.

---

<sub>Also, maybe we could change `.sublimeformatter.user` to `.sublimeformatter-user`? Don't want Windows thinking that `*.user` is an actual file extension.</sub>

**Comment by bitst0rm** [comment]
@midrare 
Thanks for your suggestions and explanations 👍
I have completed the auto-format feature and released it as 1.4.1, the following changes:
- Rename: .sublimeformatter.user to .sublimeformatter-user
- Add: Merging maps
- Add: Overwriting parent map with child map (bottom-up override)
- auto_format JSON in `Formatter.sublime-settings`
- so outside and inside are completed with just one `Auto Format File`

**Comment by midrare** [comment]
@bitst0rm Tested on Windows, everything is working for me. Thanks for the quick work 💯

**Comment by bitst0rm** [comment]
@midrare Thank you for all your help and feedback!
I'll reach out to you when I have questions that need your insightful advice.
